client/src/main/java/io/pravega/client/connection/impl/CommandEncoder.java

```diff
@@ -0,0 +1,419 @@
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.shared.NameUtils.segmentTags;
+import static io.pravega.shared.metrics.ClientMetricKeys.CLIENT_APPEND_BLOCK_SIZE;
+import static io.pravega.shared.protocol.netty.WireCommands.TYPE_PLUS_LENGTH_SIZE;
+import static io.pravega.shared.protocol.netty.WireCommands.TYPE_SIZE;
+
+@Slf4j
+@RequiredArgsConstructor
+public class CommandEncoder {
+    
+    private static final byte[] LENGTH_PLACEHOLDER = new byte[4];
+    private final Function<Long, AppendBatchSizeTracker> appendTracker;
+    private final MetricNotifier metricNotifier;
+    @GuardedBy("$lock")
+    private final Map<Map.Entry<String, UUID>, Session> setupSegments = new HashMap<>();
+    private final AtomicLong tokenCounter = new AtomicLong(0);
+    @GuardedBy("$lock")
+    private String segmentBeingAppendedTo;
+    @GuardedBy("$lock")
+    private UUID writerIdPerformingAppends;
+    @GuardedBy("$lock")
+    private int currentBlockSize;
+    @GuardedBy("$lock")
+    private int bytesLeftInBlock;
+    @GuardedBy("$lock")
+    private final Map<UUID, Session> pendingWrites = new HashMap<>();
+
+    private final OutputStream output;
+    private final ByteBuf buffer = Unpooled.buffer(1024 * 1024);
+
+    @RequiredArgsConstructor
+    private final class Session {
+        private static final int MAX_EVENTS = 500;
+        private static final int MAX_DATA_SIZE = 1024 * 1024; // 1MB
+        private final UUID id;
+        private final long requestId;
+        private final List<ByteBuf> pendingList = new ArrayList<>();
+        private int  pendingBytes = 0;
+        private long lastEventNumber = -1L;
+        private int eventCount = 0;
+
+        /**
+         * Record the given Append to Session by tracking the last event number and number of events to write.
+         */
+        private void record(Append append) {
+            lastEventNumber = append.getEventNumber();
+            eventCount += append.getEventCount();
+        }
+
+        /**
+         * Check if session is empty or not.
+         * @return true if there are any pending events; false otherwise.
+         */
+        private boolean isFree() {
+            return eventCount == 0;
+        }
+
+        /**
+         * Queue the Append data to Session' list.
+         *
+         * @param data  data bytes.
+         * @param out   Network channel buffer.
+         */
+        private void write(ByteBuf data) {
+            pendingWrites.putIfAbsent(id, this);
```

Comment:
I don't want to special case that too much. It is not outlawed by the wireprotocol to append zero bytes. (I don't have a reason we should, but we did and the wire command encoding layer is 'smart' and optimizes out the write it might be confusing)